docs/Mandoline.1
.\"Modified from man(1) of FreeBSD, the NetBSD mdoc.template, and mdoc.samples.
.\"See Also:
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd 10/13/10               \" DATE 
.Dt Mandoline 1      \" Program name and manual section number 
.Os Darwin
.Sh NAME                 \" Section Header - required - don't modify 
.Nm Mandoline
.\" The following lines are read in generating the apropos(man -k) database. Use only key
.\" words here as the database is built based on the words here and in the .ND line. 
.\" .Nm Other_name_for_same_program(),
.\" .Nm Yet another name for the same program.
.\" Use .Nm macro to designate other names for the documented program.
.Nd STL file slicer to generate gcode files for 3D FDM printing.
.Sh SYNOPSIS             \" Section Header - required - don't modify
.Nm
.Op Fl c                 \" [-c]
.Op Fl d Ar PREFIX       \" [-d PREFIX] 
.Op Fl f Ar FLOAT        \" [-f FLOAT] 
.Op Fl F Ar FLOAT        \" [-F FLOAT] 
.Op Fl i Ar FLOAT        \" [-i FLOAT] 
.Op Fl l Ar FLOAT        \" [-l FLOAT] 
.Op Fl m Ar STRING       \" [-m STRING] 
.Op Fl o Ar FILENAME     \" [-o FILENAME] 
.Op Fl p Ar INT          \" [-p INT] 
.Op Fl r Ar FLOAT        \" [-r FLOAT] 
.Op Fl s Ar FLOAT        \" [-s FLOAT] 
.Op Fl z Ar FLOAT        \" [-z path] 
.Ar FILE                 \" [file]
.\" .Op Ar                   \" [file ...]
.\" .Ar arg0                 \" Underlined argument - use .Ar anywhere to underline
.\" arg2 ...                 \" Arguments
.Sh DESCRIPTION          \" Section Header - required - don't modify
.Nm
is used to slice an STL file into the tool paths needed by
FDM type 3D printers like RepRap or MakerBot printers.
.\" Underlining is accomplished with the .Ar macro like this:
.\" .Ar underlined text .
.\" .Pp                      \" Inserts a space
.\" A list of items with descriptions:
.\" .Bl -tag -width -indent  \" Begins a tagged list 
.\" .It item a               \" Each item preceded by .It macro
.\" Description of item a
.\" .It item b
.\" Description of item b
.\" .El                      \" Ends the list
.Pp
A list of flags and their descriptions:
.Bl -tag -width -indent  \" Differs from above in tag removed 
.It Fl c                 \"-a flag as a list item
Turns off automatic centering and placing on platform of the model.
.It Fl d Ar PREFIX
Causes each slice layer to be dumped to an SVG file with the given prefix.
.It Fl f Ar FLOAT
Input filament diameter in millimeters.
.It Fl F Ar FLOAT
Input filament feedrate in millimeters per second.
.It Fl i Ar FLOAT
Infill density.  0.0 to 1.0.
.It Fl l Ar FLOAT
Slice layer thickness, in millimeters.  (0.25 to 0.4 recommended.)
.It Fl m Ar STRING
Extruded material type to get default settings for. (Usually ABS.)
.It Fl o Ar FILENAME
Name of file to write output to.  Usually goes to stdout.
.It Fl p Ar INT
Number of perimeter shells to create. (1 to 3 recommended.)
.It Fl r Ar FLOAT
Extrusion width over thickness ratio. (1.2 to 2.0 recommended.)
.It Fl s Ar FLOAT
Scale model by the given factor.
.It Fl z Ar FLOAT
Slice only the layer at the given Z level.  Useful with -d for debugging.
.El                      \" Ends the list
.Pp
.\" .Sh ENVIRONMENT      \" May not be needed
.\" .Bl -tag -width "ENV_VAR_1" -indent \" ENV_VAR_1 is width of the string ENV_VAR_1
.\" .It Ev ENV_VAR_1
.\" Description of ENV_VAR_1
.\" .It Ev ENV_VAR_2
.\" Description of ENV_VAR_2
.\" .El                      
.Sh FILES                \" File used or created by the topic of the man page
.Bl -tag -width "~/.mandoline/materials" -compact
.It Pa ~/.mandoline/defaults
The defaults for all the
.Nm
settings.
.It Pa ~/.mandoline/materials
Material specific settings for
.Nm .
.El                      \" Ends the list
.\" .Sh DIAGNOSTICS       \" May not be needed
.\" .Bl -diag
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .El
.Sh SEE ALSO 
.\" List links in ascending order by section, alphabetically within a section.
.\" Please do not reference files that do not exist without filing a bug report
.Xr skeinforge.py 1 , 
.Xr ReplicatorG 1
.Sh BUGS              \" Document known, unremedied bugs 
Numerous.
Can't read ASCII STL files.
.\" .Sh HISTORY           \" Document history if command behaves in a unique manner


src/Mandoline.cc
#include <unistd.h>
#include <getopt.h>
#include "Defaults.h"
#include "Stopwatch.h"
#include "SlicingContext.h"
#include "Operation.h"
#include "OpQueue.h"
#include "CarveOp.h"
#include "InfillOp.h"
#include "InsetOp.h"
#include "SvgDumpOp.h"
#include "PathFinderOp.h"
#include "GCodeExportOp.h"
#include "BGL/BGL.h"

static string inFileName  = "";
static string material    = "ABS";
static float scaling      = 1.0f;
static float rotation     = 0.0f;
static bool  doCenter     = true;
static float onlyAtZ      = -1.0;
static bool  doDumpSVG    = false;
static int   threadcount  = DEFAULT_WORKER_THREADS;

enum ExportTypes {
    NONE,
    GCODE,
    S3G
} exportType = GCODE;



void usage(const char* arg0, SlicingContext& ctx)
{
    fprintf(stderr, "Usage: %s [OPTIONS] FILE\n", arg0);
    fprintf(stderr, "Or   : %s -m MATERIAL [OPTIONS] FILE\n", arg0);
    fprintf(stderr, "\t[-m STRING]   Extruded material. (default ABS)\n");
    fprintf(stderr, "\t[-f FLOAT]    Filament diameter. (default %.1f mm)\n", ctx.filamentDiameter);
    fprintf(stderr, "\t[-F FLOAT]    Filament feedrate. (default %.3f mm/s)\n", ctx.filamentFeedRate);
    fprintf(stderr, "\t[-i FLOAT]    Infill density. (default %.2f)\n", ctx.infillDensity);
    fprintf(stderr, "\t[-l FLOAT]    Slicing layer thickness. (default %.2f mm)\n", ctx.layerThickness);
    fprintf(stderr, "\t[-p INT]      Number of perimeter shell layers. (default %d)\n", ctx.perimeterShells);
    fprintf(stderr, "\t[-w FLOAT]    Extrusion width over thickness ratio. (default %.2f)\n", ctx.widthOverHeightRatio);
    fprintf(stderr, "\t[-c]          DON'T center model on platform before slicing.\n");
    fprintf(stderr, "\t[-s FLOAT]    Scale model.  (default %.4gx)\n", scaling);
    fprintf(stderr, "\t[-r FLOAT]    Rotate model about Z.  (default %.4g deg)\n", rotation);
    fprintf(stderr, "\t[-z FLOAT]    Slice model only at the given Z level.\n");
    fprintf(stderr, "\t[-d PREFIX]   Dump layers to SVG files with names like PREFIX-12.34.svg.\n");
    fprintf(stderr, "\t[-t INT]      Number of threads to slice with. (default %d)\n", threadcount);
    exit(-1);
}



int main (int argc, char * const argv[])
{
    Stopwatch stopwatch;
    stopwatch.start();
    
    SlicingContext ctx;
    char buf[512];

    int ch;
    const char *progName = argv[0];
    const char * shortopts = "?cd:f:F:hi:l:m:o:p:r:s:t:w:Z:";
    static struct option longopts[] = {
	{"material", required_argument, NULL, 'm'},
	{"diameter", required_argument, NULL, 'f'},
	{"feedrate", required_argument, NULL, 'F'},
	{"infill", required_argument, NULL, 'i'},
	{"layer", required_argument, NULL, 'l'},
	{"shells", required_argument, NULL, 'p'},
	{"ratio", required_argument, NULL, 'w'},
	{"nocenter", required_argument, NULL, 'c'},
	{"scale", required_argument, NULL, 's'},
	{"rotatex", required_argument, NULL, 'r'},
	{"onlyatz", required_argument, NULL, 'Z'},
	{"dumpprefix", required_argument, NULL, 'd'},
	{"threads", required_argument, NULL, 't'},
	{0, 0, 0, 0}
    };
    
    // We need to get the command-line argument for Material first.
    while ((ch = getopt_long(argc, argv, shortopts, longopts, NULL)) != -1) {
        if (ch == 'm') {
	    material = optarg;
	}
    }

    // Load saved settings for the given material, if any.
    snprintf(buf, sizeof(buf), "~/.mandoline/materials/%.128s.conf", material.c_str());
    ctx.loadDefaultsFromFile(buf);

    // Process remainder of command-line arguments.
    optreset = opterr = optind = 1;
    while ((ch = getopt_long(argc, argv, shortopts, longopts, NULL)) != -1) {
        switch (ch) {
            case 'c':
                doCenter = false;
                break;
            case 'd':
                doDumpSVG = true;
                ctx.dumpPrefix = optarg;
                break;
            case 'f':
                ctx.filamentDiameter = atof(optarg);
                break;
            case 'F':
                ctx.filamentFeedRate = atof(optarg);
                break;
            case 'i':
                ctx.infillDensity = atof(optarg);
                break;
            case 'l':
                ctx.layerThickness = atof(optarg);
                break;
            case 'p':
                ctx.perimeterShells = atoi(optarg);
                break;
            case 'w':
                ctx.widthOverHeightRatio = atof(optarg);
                break;
            case 'r':
                rotation = atof(optarg);
                break;
            case 's':
                scaling = atof(optarg);
                break;
            case 't': {
		    int cnt = atoi(optarg);
		    if (cnt < 1) {
			fprintf(stderr, "Error: Thread count cannot be less than 1.\n");
		    }
		    threadcount = cnt;
		}
                break;
            case 'Z':
                onlyAtZ = atof(optarg);
                break;
            case 'm':
		// We already parsed this one out.  Ignore.
                break;
            case '?':
            case 'h':
            default:
                usage(progName, ctx);
        }
    }
    argc -= optind;
    argv += optind;
    if (argc > 0) {
        inFileName = string(argv[0]);
    }
    if (inFileName.length() < 1) {
        usage(progName, ctx);
    }
    
    // Load the model from the file.
    BGL::Mesh3d &mesh = ctx.mesh;
    mesh.loadFromSTLFile(inFileName.c_str());
    printf("Found %d faces.\n", mesh.size());
    stopwatch.checkpoint("Model loaded from file");
    printf("Model Bounds = (%.2f, %.2f, %.2f) to (%.2f, %.2f, %.2f)\n", mesh.minX, mesh.minY, mesh.minZ, mesh.maxX, mesh.maxY, mesh.maxZ);
    
    // Scale the model if requested.
    if (scaling != 1.0f) {
        printf("Scaling model by %.4gx\n", scaling);
        mesh.scale(scaling);
    }

    // Rotate the model if requested.
    if (rotation != 0.0f) {
        printf("Rotating model by %.4g degrees\n", rotation);
        mesh.rotateZ(rotation*M_PI/180.0f);
    }

    // Center the model.
    if (doCenter) {
        printf("Centering model on X=0, Y=0  Placing bottom at Z=0.\n");
        mesh.translateToCenterOfPlatform();
    }    
    
    if (scaling != 1.0f || rotation != 0.0f || doCenter) {
	printf("New Bounds = (%.2f, %.2f, %.2f) to (%.2f, %.2f, %.2f)\n", mesh.minX, mesh.minY, mesh.minZ, mesh.maxX, mesh.maxY, mesh.maxZ);
        stopwatch.checkpoint("Transformed");
    }
    
    // Recalculate SVG offsets.
    ctx.calculateSvgOffsets();
    
    // Calculate first and last layer Zs
    printf("Layer Thickness=%.4g\n", ctx.layerThickness);
    float z = ctx.layerThickness/2.0f;
    float topZ = mesh.maxZ;
    if (onlyAtZ > 0.0f) {
        z = onlyAtZ;
        topZ = onlyAtZ + ctx.layerThickness / 2.0f;
    }
    
    // Set up Operations Queue and threadpool.
    OpQueue opQ;
    opQ.setMaxConcurrentOperationCount(threadcount);

    // Carve model to find layer outlines
    while (z < topZ) {
        CarveOp* op = new CarveOp(&ctx, ctx.allocSlice(z), z);
	opQ.addOperation(op);
        z += ctx.layerThickness;
    }
    opQ.waitUntilAllOperationsAreFinished();
    stopwatch.checkpoint("Carved");
    
    map<float,CarvedSlice>::iterator it;

    // Inset each level's carved region
    for (it = ctx.slices.begin(); it != ctx.slices.end(); it++) {
        InsetOp* op = new InsetOp(&ctx, &(*it).second, (*it).first);
	opQ.addOperation(op);
    }
    opQ.waitUntilAllOperationsAreFinished();
    stopwatch.checkpoint("Inset");
    
    // Infill each level's carved region
    for (it = ctx.slices.begin(); it != ctx.slices.end(); it++) {
        InfillOp* op = new InfillOp(&ctx, &(*it).second, (*it).first);
	opQ.addOperation(op);
    }
    opQ.waitUntilAllOperationsAreFinished();
    stopwatch.checkpoint("Infilled");
    
    // Optionally dump to SVG
    if (doDumpSVG) {
	for (it = ctx.slices.begin(); it != ctx.slices.end(); it++) {
	    SvgDumpOp* op = new SvgDumpOp(&ctx, &(*it).second, (*it).first);
	    opQ.addOperation(op);
        }
	opQ.waitUntilAllOperationsAreFinished();
        stopwatch.checkpoint("Dumped to SVG");
    }

    // Find optimized path.
    PathFinderOp* op = new PathFinderOp(&ctx);
    opQ.addOperation(op);
    opQ.waitUntilAllOperationsAreFinished();
    stopwatch.checkpoint("Path Optimized");
    
    // Export toolpaths to apropriate output format.
    Operation* expOp = NULL;
    switch (exportType) {
        case GCODE:
	    expOp = new GCodeExportOp(&ctx);
	    break;
	default:
	    break;
    }
    if (expOp) {
	opQ.addOperation(expOp);
	opQ.waitUntilAllOperationsAreFinished();
	stopwatch.checkpoint("Exported");
    }

    // All done!
    stopwatch.finish();
    
    return 0;
}

